src/backend/columnar/columnar_tableam.c

```diff
@@ -1178,24 +1201,83 @@ columnar_index_build_range_scan(Relation columnarRelation,
 }
 
 
+/*
+ * ColumnarGetNumberOfVirtualBlocks returns total number of "virtual" blocks
+ * that given columnar table has based on based on ItemPointer BlockNumber's.
+ */
+static BlockNumber
+ColumnarGetNumberOfVirtualBlocks(Relation relation, Snapshot snapshot)
+{
+	ItemPointerData highestItemPointer =
+		ColumnarGetHighestItemPointer(relation, snapshot);
+	if (!ItemPointerIsValid(&highestItemPointer))
+	{
+		/* table is empty according to our snapshot */
+		return 0;
+	}
+
+	/*
+	 * Since BlockNumber is 0-based, increment it by 1 to find the total
+	 * number of "virtual" blocks.
+	 */
+	return ItemPointerGetBlockNumber(&highestItemPointer) + 1;
+}
+
+
+/*
+ * ColumnarGetHighestItemPointer returns ItemPointerData for the tuple with
+ * highest tid for given relation.
+ * If given relation is empty, then returns invalid item pointer.
+ */
+static ItemPointerData
+ColumnarGetHighestItemPointer(Relation relation, Snapshot snapshot)
+{
+	StripeMetadata *stripeWithHighestRowNumber =
+		FindStripeWithHighestRowNumber(relation, snapshot);
```

Comment:
how important is it we return the actual committed highest tid?
can't we simply read the reservation from the metapage and use that, compared to scanning an index backwards?

Reply:
If we consider uncommited highest row number, then progress reporting might be little confusing.
How we report the progress is like:
1) Report total number of blocks *to* read by using itemPointer->block for the highest row number
2) For each tuple, report the block itemPointer->block as read
3) Report total number of blocks *as* read by using itemPointer->block for the same highest row number

Here, if we consider uncommitted reservations in steps 1 & 3, then we would have a jump from last reported block number (2)) to the block number of the reservation made for the uncommitted transaction.